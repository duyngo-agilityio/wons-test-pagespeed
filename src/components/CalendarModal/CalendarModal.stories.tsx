import { useState } from 'react';

import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

// Components
import CalendarModal from './index';
import { Button } from '../common';

const meta: Meta<typeof CalendarModal> = {
  title: 'Components/CalendarModal',
  component: CalendarModal,
  parameters: {
    layout: 'centered',
    controls: { expanded: true },
  },
};

export default meta;
type Story = StoryObj<typeof CalendarModal>;

const mockTimeRange = {
  start: '12:00am',
  end: '12:30am',
};

const BasicUsage = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <>
      <Button
        className="mt-4"
        color="primary"
        type="button"
        onClick={() => setIsOpen(true)}
      >
        Open modal
      </Button>

      <CalendarModal
        title="Create Event"
        eventTitle="Example title"
        date={new Date()}
        timeRange={mockTimeRange}
        isOpen={isOpen}
        onSubmit={fn()}
        onClose={() => setIsOpen(false)}
      />
    </>
  );
};

export const Default: Story = {
  render: () => <BasicUsage />,
};
