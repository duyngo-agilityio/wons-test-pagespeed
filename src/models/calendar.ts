export interface IEvent {
  id: number;
  title: string;
  status: string;
  location: string;
  visibility: string;
  date: Date;
  startTime: string;
  endTime: string;
  eventType: string;
  repeatSetting: string;
  notificationTime: string;
  users_permissions_users: number[];
}

export interface ICalendarTask {
  id: number;
  title: string;
  date: Date;
  time: string;
  descriptions: string;
}
