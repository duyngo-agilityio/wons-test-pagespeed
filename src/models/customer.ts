type TGender = 'male' | 'female';

export interface ICustomer {
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  gender: TGender;
  job: string;
  address: string;
  avatar: string;
}
