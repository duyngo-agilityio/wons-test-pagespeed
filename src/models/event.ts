export interface IEvent {
  id: number;
  title: string;
  status: string;
  location: string;
  visibility: string;
  date: Date;
  startTime: string;
  endTime: string;
  eventType: string;
  repeatSetting: string;
  notificationTime: string;
  users_permissions_users: number[];
}
