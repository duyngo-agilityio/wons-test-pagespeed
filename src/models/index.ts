export * from './customer';
export * from './invoice';
export * from './product';
export * from './customer';
export * from './invoice';
export * from './statistic';
export * from './user';
export * from './calendar';
