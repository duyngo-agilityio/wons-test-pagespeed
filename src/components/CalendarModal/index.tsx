'use client';

import { memo, useState } from 'react';
import { CalendarDate } from '@internationalized/date';
import { IoClose } from 'react-icons/io5';
import isEqual from 'react-fast-compare';

import { Modal as NextModal, ModalContent, Calendar } from '@nextui-org/react';

// Utils
import { formatToCalendarDate } from '@/utils';

// Models
import { IEvent, TUser } from '@/models';

// Components
import {
  Heading,
  Button,
  Tabs,
  CalendarEventForm,
  CalendarTaskForm,
} from '@/components';

interface TimeRangeProps {
  start: string;
  end: string;
}

interface EventForm {
  title: string;
  location?: string;
  people?: string | number[];
}

interface TaskForm {
  title: string;
  description: string;
}

interface EventFormModalProps {
  title: string;
  eventTitle: string;
  date: Date;
  timeRange?: TimeRangeProps;
  time?: string;
  isOpen: boolean;
  repeatSetting?: string;
  guests?: string[];
  location?: string;
  user?: TUser;
  previewData?: EventForm | TaskForm | null;
  isEdit?: boolean;
  isTask?: boolean;
  setIsTask?: (value: boolean) => void;
  onSubmit: (data: Partial<IEvent>) => void;
  onClose: () => void;
}

const CalendarModal = ({
  title = '',
  eventTitle = '',
  date,
  timeRange,
  isEdit = false,
  isTask = false,
  time,
  isOpen,
  user,
  previewData = null,
  setIsTask,
  onClose,
  onSubmit,
}: EventFormModalProps): JSX.Element => {
  const [isOpenCalendar, setIsOpenCalendar] = useState(false);
  const [calendarDate, setCalendarDate] = useState(formatToCalendarDate(date));

  const handleDateChange = (newDate: CalendarDate) => {
    setIsOpenCalendar(false);
    setCalendarDate(newDate);
  };

  const handleModalClose = () => {
    onClose();
  };

  const handleSelectionChange = (key: React.Key) => {
    if (key === 'task') {
      setIsTask && setIsTask(true);
    }
  };

  const calendarTabs = [
    {
      key: 'event',
      label: 'Event',
      content: (
        <CalendarEventForm
          previewData={previewData}
          timeRange={timeRange}
          eventTitle={eventTitle}
          date={date}
          user={user}
          onSubmit={onSubmit}
          onClose={onClose}
        />
      ),
      isDisable: isEdit && isTask,
    },
    {
      key: 'task',
      label: 'Task',
      content: (
        <CalendarTaskForm
          previewData={previewData}
          date={date}
          onClose={onClose}
          time={time ?? ''}
          onSubmit={onSubmit}
        />
      ),
      isDisable: isEdit && !isTask,
    },
    {
      key: 'reminder',
      label: 'Reminder',
      content: 'This is the reminder content',
      isDisable: true,
    },
  ];

  return (
    <NextModal
      className="max-w-[467px]"
      isOpen={isOpen}
      size="xl"
      hideCloseButton={true}
      placement="center"
    >
      <ModalContent className="relative top-0 left-0 p-[30px_30px_40px] bg-white dark:bg-gray-800 rounded-[10px] shadow-[-14px_30px_20px_0px_rgba(0,0,0,0.05)] w-[467px]">
        {isOpenCalendar && (
          <Calendar
            className="absolute top-[240px] left-[80px] z-10"
            aria-label="Date"
            value={calendarDate}
            onChange={handleDateChange}
          />
        )}

        <Heading className="mt-[12px]" size="md" title={title} />

        <Button
          className="absolute top-[30px] right-[30px] bg-pink-50 dark:bg-pink-600 text-pink-500 dark:text-pink-500 border-none rounded-full w-10 h-10 flex justify-center items-center cursor-pointer px-0"
          data-testid="close-button"
          onClick={handleModalClose}
        >
          <IoClose size={20} />
        </Button>

        <Tabs
          style={{ padding: 0, margin: '40px 0 18px' }}
          tabs={calendarTabs}
          defaultSelectedKey={isEdit && isTask ? 'task' : 'event'}
          onSelectionChange={handleSelectionChange}
        />
      </ModalContent>
    </NextModal>
  );
};

export default memo(CalendarModal, isEqual);
