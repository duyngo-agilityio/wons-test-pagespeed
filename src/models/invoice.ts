// Types
import { ICustomer } from './customer';

// Constants
import { InvoiceStatus } from '@/constants';

export type TInvoiceProduct<T> = {
  price: number;
  createdAt: string;
  updatedAt: string;
  publishedAt: string;
  quantity: number;
  product: { data: T };
};

export interface IInvoice {
  id: string;
  customer: ICustomer;
  imageUrl: string;
  status: InvoiceStatus;
  address: string;
  isSelected: boolean;
  date: string;
  email: string;
}
