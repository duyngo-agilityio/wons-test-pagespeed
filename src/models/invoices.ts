// Constants
import { InvoiceStatus } from '@/constants';

// Types
import { ICustomer } from './customer';

export interface IInvoice {
  id: string;
  email: string;
  imageUrl: string;
  customer: ICustomer;
  date: string;
  status: InvoiceStatus;
  address: string;
  isSelected: boolean;
}
