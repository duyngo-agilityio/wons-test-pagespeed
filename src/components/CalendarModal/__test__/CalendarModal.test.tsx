import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

// Components
import CalendarModal from '../index';

// Models
import { TUser } from '@/models';

describe('CalendarModal Component', () => {
  const mockOnClose = jest.fn();
  const mockOnSubmit = jest.fn();
  const mockSetIsTask = jest.fn();

  const defaultProps = {
    title: 'Calendar Modal',
    eventTitle: 'Sample Event',
    date: new Date(),
    timeRange: { start: '10:00', end: '11:00' },
    time: undefined,
    isOpen: true,
    user: {} as TUser,
    previewData: null,
    isTask: false,
    isEdit: false,
    setIsTask: mockSetIsTask,
    onClose: mockOnClose,
    onSubmit: mockOnSubmit,
  };

  const setup = (props = defaultProps) => render(<CalendarModal {...props} />);
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders modal with correct title and close button', () => {
    setup();
    expect(screen.getByText(defaultProps.title)).toBeInTheDocument();
    expect(screen.getByTestId('close-button')).toBeInTheDocument();
  });

  it('calls onClose when the close button is clicked', () => {
    setup();
    const closeButton = screen.getByTestId('close-button');
    fireEvent.click(closeButton);
    expect(mockOnClose).toHaveBeenCalled();
  });

  it('renders CalendarTaskForm when "Task" tab is selected', () => {
    setup({ ...defaultProps, isEdit: false, isTask: true });
    const taskTab = screen.getByText('Task');
    fireEvent.click(taskTab);
    expect(screen.getByText('Task')).toBeInTheDocument();
  });

  it('disables the "Task" tab when isEdit is true and isTask is false', () => {
    setup({ ...defaultProps, isEdit: true, isTask: false });
    const taskTab = screen.getByText('Task');
    expect(taskTab).toHaveClass(
      'relative z-10 whitespace-nowrap transition-colors text-gray-700 dark:text-pink-700 group-data-[selected=true]:text-white',
    );
  });

  it('calls setIsTask when "Task" tab is selected', () => {
    setup();
    const taskTab = screen.getByText('Task');
    fireEvent.click(taskTab);
    expect(mockSetIsTask).toHaveBeenCalledWith(true);
  });

  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
});
