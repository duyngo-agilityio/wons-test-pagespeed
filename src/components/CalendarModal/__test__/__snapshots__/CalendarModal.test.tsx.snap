// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarModal Component matches snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;
